File: app/csharp-ms/src/ManagementAccounting/ManagementAccounting.Infrastructure/Adapters/External/FinancialAccountingClient.cs
using ManagementAccounting.Application.Ports.Out;
using ManagementAccounting.Domain.Entities;

namespace ManagementAccounting.Infrastructure.Adapters.External;

/// <summary>
/// 財務会計サービスクライアント（腐敗防止層）
///
/// 財務会計コンテキストのデータモデルの変更から
/// 管理会計コンテキストを保護する
/// </summary>
public class FinancialAccountingClient : IFinancialAccountingClient
{
    private readonly IFinancialAccountingApi _api;

    public FinancialAccountingClient(IFinancialAccountingApi api)
    {
        _api = api;
    }

    /// <summary>
    /// 財務会計サービスからデータを取得し、管理会計ドメインモデルに変換
    /// </summary>
    public async Task<FinancialData> FetchFinancialDataByFiscalYearAsync(int fiscalYear)
    {
        var journals = await _api.GetJournalsByFiscalYearAsync(fiscalYear);

        // 全仕訳明細をフラット化
        var entries = journals.SelectMany(j => j.Entries).ToList();

        // 管理会計コンテキストのドメインモデルに変換
        return ConvertToFinancialData(fiscalYear, entries);
    }

    /// <summary>
    /// 財務会計のデータモデルを管理会計のドメインモデルに変換
    ///
    /// この変換ロジックにより、財務会計のスキーマ変更の影響を局所化
    /// </summary>
    private FinancialData ConvertToFinancialData(int fiscalYear, List<JournalEntryDto> entries)
    {
        // 売上高（41 から始まる勘定科目）
        var sales = SumByAccountPrefix(entries, "41", "credit");

        // 営業利益（簡易計算: 収益 - 費用）
        var revenue = SumByAccountPrefix(entries, "41", "credit");
        var costOfSales = SumByAccountPrefix(entries, "51", "debit");
        var sellingExpenses = SumByAccountPrefix(entries, "52", "debit");
        var operatingProfit = revenue - costOfSales - sellingExpenses;

        // 総資産（1 から始まる勘定科目）
        var totalAssets = SumByAccountPrefix(entries, "1", "debit");

        // 有形固定資産（121 から始まる勘定科目）
        var tangibleFixedAssets = SumByAccountPrefix(entries, "121", "debit");

        // 流動資産（11 から始まる勘定科目）
        var currentAssets = SumByAccountPrefix(entries, "11", "debit");

        // 流動負債（21 から始まる勘定科目）
        var currentLiabilities = SumByAccountPrefix(entries, "21", "credit");

        // 当座資産（現金預金 + 売掛金）
        var quickAssets = SumByAccountCode(entries, "111", "debit")   // 現金預金
                        + SumByAccountCode(entries, "112", "debit");  // 売掛金

        // 自己資本（3 から始まる勘定科目）
        var equity = SumByAccountPrefix(entries, "3", "credit");

        return new FinancialData(
            FiscalYear: fiscalYear,
            Sales: sales,
            OperatingProfit: operatingProfit,
            TotalAssets: totalAssets,
            TangibleFixedAssets: tangibleFixedAssets,
            CurrentAssets: currentAssets,
            CurrentLiabilities: currentLiabilities,
            QuickAssets: quickAssets,
            Equity: equity
        );
    }

    private decimal SumByAccountPrefix(List<JournalEntryDto> entries, string prefix, string side)
    {
        return entries
            .Where(e => e.AccountCode.StartsWith(prefix))
            .Sum(e => side == "debit" ? e.DebitAmount : e.CreditAmount);
    }

    private decimal SumByAccountCode(List<JournalEntryDto> entries, string accountCode, string side)
    {
        return entries
            .Where(e => e.AccountCode == accountCode)
            .Sum(e => side == "debit" ? e.DebitAmount : e.CreditAmount);
    }
}

File: app/fsharp-ms/src/ManagementAccounting.Infrastructure/Adapters/External/FinancialAccountingClient.fs
namespace ManagementAccounting.Infrastructure.Adapters.External

open System
open System.Net.Http
open System.Net.Http.Json
open System.Threading.Tasks
open ManagementAccounting.Domain.Models
open ManagementAccounting.Application.Ports.Out

/// <summary>
/// 財務会計サービスから取得する仕訳 DTO
/// </summary>
[<CLIMutable>]
type JournalDto = {
    JournalId: int
    JournalDate: DateTime
    Description: string
    FiscalYear: int
}

/// <summary>
/// 財務会計サービスから取得する仕訳明細 DTO
/// </summary>
[<CLIMutable>]
type JournalEntryDto = {
    AccountCode: string
    DebitAmount: decimal
    CreditAmount: decimal
    Description: string
}

/// <summary>
/// 財務会計サービスから取得する勘定科目 DTO
/// </summary>
[<CLIMutable>]
type AccountDto = {
    AccountId: int
    AccountCode: string
    AccountName: string
    AccountType: string
    Balance: decimal
}

/// <summary>
/// 財務会計サービスクライアント（腐敗防止層 - ACL）
/// </summary>
type FinancialAccountingClient(httpClient: HttpClient) =

    let convertToFinancialData (fiscalYear: int) (accounts: AccountDto list) : FinancialData =
        // 勘定科目種別ごとに残高を集計
        let getBalanceByType accountType =
            accounts
            |> List.filter (fun a -> a.AccountType = accountType)
            |> List.sumBy (fun a -> a.Balance)

        let getBalanceByCode code =
            accounts
            |> List.tryFind (fun a -> a.AccountCode = code)
            |> Option.map (fun a -> a.Balance)
            |> Option.defaultValue 0m

        // 資産（流動資産 + 固定資産）
        let assets = getBalanceByType "Asset"

        // 負債
        let liabilities = getBalanceByType "Liability"

        // 純資産
        let equity = getBalanceByType "Equity"

        // 収益（売上高）
        let revenue = getBalanceByType "Revenue"

        // 費用
        let expenses = getBalanceByType "Expense"

        // 営業利益 = 売上高 - 費用（簡易計算）
        let operatingProfit = revenue - expenses

        // 流動資産（勘定科目コード 1xxx）
        let currentAssets =
            accounts
            |> List.filter (fun a -> a.AccountType = "Asset" && a.AccountCode.StartsWith("1"))
            |> List.sumBy (fun a -> a.Balance)

        // 固定資産（勘定科目コード 2xxx）
        let fixedAssets =
            accounts
            |> List.filter (fun a -> a.AccountType = "Asset" && a.AccountCode.StartsWith("2"))
            |> List.sumBy (fun a -> a.Balance)

        // 流動負債
        let currentLiabilities =
            accounts
            |> List.filter (fun a -> a.AccountType = "Liability" && a.AccountCode.StartsWith("3"))
            |> List.sumBy (fun a -> a.Balance)

        // 当座資産（現金預金 + 売掛金）
        let quickAssets =
            accounts
            |> List.filter (fun a ->
                a.AccountCode = "1110" || // 現金預金
                a.AccountCode = "1120")   // 売掛金
            |> List.sumBy (fun a -> a.Balance)

        {
            FiscalYear = fiscalYear
            Sales = revenue
            OperatingProfit = operatingProfit
            TotalAssets = assets
            TangibleFixedAssets = fixedAssets
            CurrentAssets = currentAssets
            CurrentLiabilities = if currentLiabilities = 0m then liabilities else currentLiabilities
            QuickAssets = quickAssets
            Equity = equity
        }

    interface IFinancialDataPort with
        member _.FetchFinancialDataByFiscalYearAsync(fiscalYear: int) : Task<FinancialData> =
            task {
                // 財務会計サービスから勘定科目一覧を取得
                let! response = httpClient.GetAsync("/api/accounts")
                response.EnsureSuccessStatusCode() |> ignore

                let! accounts = response.Content.ReadFromJsonAsync<AccountDto list>()
                let accountList = accounts |> Option.ofObj |> Option.defaultValue []

                return convertToFinancialData fiscalYear accountList
            }
